# NORT Inference — Apply Trained Classifier to NOVEL Videos

This notebook applies a trained SVM classifier to all NOVEL-phase NORT videos to produce per-frame exploration predictions.

**Inputs:**
- `svm_exploratory_classifier.pkl` — trained RBF-SVM
- `feature_scaler.pkl` — fitted StandardScaler from training
- NOVEL-phase `.mp4` videos + their DLC `.csv` files
- Per-video `_bbox_assignments.pkl` files (bounding box positions for each frame)

**Pipeline overview:**
1. Load the trained model and scaler
2. For each NOVEL video, extract the same geometric features used during training
3. Predict exploratory / non-exploratory for each frame
4. If exploratory, assign the exploration to the closest object by nose-to-centroid distance
5. Save per-video JSON prediction files and labeled MP4s with overlaid annotations

**Output:** Per-video `_predictions.json` files aggregated into `nort_results.csv` for use in `Correlate-MoSeq-And-Traditional-Behavior.ipynb`

In [1]:
import json
import pickle
import pandas as pd
import numpy as np
from pathlib import Path
import os
import cv2
import shutil
from tqdm import tqdm
import joblib
from sklearn.svm import SVC
from sklearn.preprocessing import StandardScaler

# Load the trained model and scaler
model_dir = "/mnt/g/Shared drives/llorente-lab/Moseq/AnalyzedData/Revision-NORT/nort_model"
model_file = os.path.join(model_dir, "svm_exploratory_classifier.pkl")
scaler_file = os.path.join(model_dir, "feature_scaler.pkl")

svm_optimized = joblib.load(model_file)
scaler = joblib.load(scaler_file)

print("Model and scaler loaded!")
print(f"Model: {type(svm_optimized)}")
print(f"Scaler: {type(scaler)}")

Model and scaler loaded!
Model: <class 'sklearn.svm._classes.SVC'>
Scaler: <class 'sklearn.preprocessing._data.StandardScaler'>


# 1. Load Trained Model

Load the RBF-SVM classifier and fitted `StandardScaler` saved by `nort.ipynb`. The scaler must be the same one used during training — applying it here ensures feature distributions match what the model was trained on.

In [2]:
def load_dlc_csv(dlc_path):
    df = pd.read_csv(dlc_path, header=[0, 1, 2], index_col=0)
    scorer = df.columns.get_level_values('scorer')[0]
    
    nose_x = df[scorer]['nose']['x'].values
    nose_y = df[scorer]['nose']['y'].values
    nose_likelihood = df[scorer]['nose']['likelihood'].values
    
    tail_x = df[scorer]['tail_base']['x'].values
    tail_y = df[scorer]['tail_base']['y'].values
    tail_likelihood = df[scorer]['tail_base']['likelihood'].values
    
    return {
        'nose_x': nose_x,
        'nose_y': nose_y,
        'nose_likelihood': nose_likelihood,
        'tail_x': tail_x,
        'tail_y': tail_y,
        'tail_likelihood': tail_likelihood
    }

def find_dlc_csv(video_path, dlc_folder):
    video_stem = Path(video_path).stem
    pattern = f"{video_stem}DLC*.csv"
    matches = [f for f in Path(dlc_folder).glob(pattern) if '_filtered' not in f.name]
    
    if len(matches) == 0:
        raise FileNotFoundError(f"No DLC CSV found for {video_stem}")
    
    return str(matches[0])

def bbox_to_vertices(bbox):
    xmin, ymin = bbox['xmin'], bbox['ymin']
    xmax, ymax = bbox['xmax'], bbox['ymax']
    
    vertices = np.array([
        [xmin, ymin],
        [xmax, ymin],
        [xmax, ymax],
        [xmin, ymax]
    ])
    return vertices

def calculate_features_for_frame(nose_x, nose_y, tail_x, tail_y, object_vertices):
    if np.isnan(nose_x) or np.isnan(nose_y) or np.isnan(tail_x) or np.isnan(tail_y):
        return np.full(8, np.nan)
    
    orientation = np.array([nose_x - tail_x, nose_y - tail_y])
    orientation_norm = np.linalg.norm(orientation)
    
    if orientation_norm < 1e-6:
        return np.full(8, np.nan)
    
    orientation = orientation / orientation_norm
    
    rotation_matrix = np.array([
        [orientation[0], orientation[1]],
        [-orientation[1], orientation[0]]
    ])
    
    features = []
    for vertex in object_vertices:
        vector = vertex - np.array([nose_x, nose_y])
        rotated = rotation_matrix @ vector
        features.extend(rotated)
    
    return np.array(features)

print("Helper functions loaded!")

Helper functions loaded!


# 2. Helper Functions

Utilities for loading DLC pose data and computing per-frame geometric features — identical to the feature engineering used during training in `nort.ipynb`.

- `load_dlc_csv` — parse DLC multi-header CSV into nose/tail arrays
- `find_dlc_csv` — locate the DLC CSV for a given video filename (handles naming variation)
- `bbox_to_vertices` — convert a `{xmin, ymin, xmax, ymax}` dict to a 4×2 vertex array
- `calculate_features_for_frame` — rotate bbox vertices into the mouse's head-direction frame and return the 8-feature vector for one object

In [3]:
def predict_with_object_assignment(video_path, svm_model, scaler, dlc_folder, bbox_assignments_dir):
    """Predict exploratory behavior per frame, then assign to closest object"""
    
    video_stem = Path(video_path).stem
    
    # Load bbox
    bbox_file = os.path.join(bbox_assignments_dir, f"{video_stem}_bbox_assignments.pkl")
    if not os.path.exists(bbox_file):
        print(f"Missing bbox for {video_stem}")
        return None
    
    with open(bbox_file, 'rb') as f:
        bbox_data = pickle.load(f)
    
    # Load DLC
    try:
        dlc_csv = find_dlc_csv(video_path, dlc_folder)
        dlc_data = load_dlc_csv(dlc_csv)
    except Exception as e:
        print(f"Error loading DLC for {video_stem}: {e}")
        return None
    
    n_frames = len(dlc_data['nose_x'])
    
    # Get object names
    first_frame_bboxes = bbox_data[0]
    object_names = sorted(first_frame_bboxes.keys())
    
    # Results
    frame_results = []
    
    for frame_idx in range(n_frames):
        if frame_idx not in bbox_data:
            frame_results.append({
                'frame_idx': frame_idx,
                'exploratory': False,
                'probability': 0.0,
                'exploring_object': None,
                'distance_to_objects': {}
            })
            continue
        
        bboxes = bbox_data[frame_idx]
        
        nose_x = dlc_data['nose_x'][frame_idx]
        nose_y = dlc_data['nose_y'][frame_idx]
        tail_x = dlc_data['tail_x'][frame_idx]
        tail_y = dlc_data['tail_y'][frame_idx]
        
        # Calculate features for ALL objects (16 features)
        all_features = []
        for obj_name in sorted(bboxes.keys()):
            vertices = bbox_to_vertices(bboxes[obj_name])
            features = calculate_features_for_frame(nose_x, nose_y, tail_x, tail_y, vertices)
            all_features.extend(features)
        
        all_features = np.array(all_features).reshape(1, -1)
        
        if np.any(np.isnan(all_features)):
            frame_results.append({
                'frame_idx': frame_idx,
                'exploratory': False,
                'probability': 0.0,
                'exploring_object': None,
                'distance_to_objects': {}
            })
            continue
        
        # Predict exploratory behavior for this frame
        features_scaled = scaler.transform(all_features)
        pred = svm_model.predict(features_scaled)[0]
        prob = svm_model.predict_proba(features_scaled)[0, 1]
        
        # Calculate distance to each object
        distances = {}
        for obj_name, bbox in bboxes.items():
            obj_center_x = (bbox['xmin'] + bbox['xmax']) / 2
            obj_center_y = (bbox['ymin'] + bbox['ymax']) / 2
            dist = np.sqrt((nose_x - obj_center_x)**2 + (nose_y - obj_center_y)**2)
            distances[obj_name] = dist
        
        # If exploratory, assign to closest object
        exploring_object = None
        if pred == 1:
            exploring_object = min(distances, key=distances.get)
        
        frame_results.append({
            'frame_idx': frame_idx,
            'exploratory': bool(pred),
            'probability': float(prob),
            'exploring_object': exploring_object,
            'distance_to_objects': distances
        })
    
    return {
        'video_stem': video_stem,
        'n_frames': n_frames,
        'object_names': object_names,
        'frame_results': frame_results
    }

def create_labeled_video(video_path, frame_results, output_path):
    """Create video with exploratory labels overlaid"""
    
    cap = cv2.VideoCapture(video_path)
    
    # Get video properties
    fps = int(cap.get(cv2.CAP_PROP_FPS))
    width = int(cap.get(cv2.CAP_PROP_FRAME_WIDTH))
    height = int(cap.get(cv2.CAP_PROP_FRAME_HEIGHT))
    
    # Create video writer
    fourcc = cv2.VideoWriter_fourcc(*'mp4v')
    out = cv2.VideoWriter(output_path, fourcc, fps, (width, height))
    
    frame_idx = 0
    while cap.isOpened():
        ret, frame = cap.read()
        if not ret:
            break
        
        if frame_idx < len(frame_results):
            result = frame_results[frame_idx]
            
            # Determine label and color
            if result['exploratory']:
                label = f"Exploring: {result['exploring_object']}"
                color = (0, 255, 0)  # Green
            else:
                label = "Non-exploratory"
                color = (0, 0, 255)  # Red
            
            # Add text to top right
            font = cv2.FONT_HERSHEY_SIMPLEX
            font_scale = 0.7
            thickness = 2
            
            # Get text size for background
            (text_width, text_height), baseline = cv2.getTextSize(label, font, font_scale, thickness)
            
            # Position in top right
            x = width - text_width - 10
            y = 30
            
            # Draw background rectangle
            cv2.rectangle(frame, (x - 5, y - text_height - 5), (x + text_width + 5, y + 5), (0, 0, 0), -1)
            
            # Draw text
            cv2.putText(frame, label, (x, y), font, font_scale, color, thickness)
        
        out.write(frame)
        frame_idx += 1
    
    cap.release()
    out.release()

print("Prediction and video functions loaded!")

Prediction and video functions loaded!


# 3. Inference Pipeline

**`predict_with_object_assignment`** — the core inference function. For each frame:
1. Load the per-frame bbox assignment and DLC coordinates
2. Compute the 16-dimensional whole-frame feature vector (both objects concatenated)
3. Predict exploratory / non-exploratory using the SVM
4. If exploratory, compute nose-to-centroid distance for each object and assign to the closest one

Returns a dict with per-frame predictions, probabilities, assigned objects, and distances.

**`create_labeled_video`** — writes a new MP4 with the prediction overlaid as text in the top-right corner (green = exploratory, red = non-exploratory).

In [4]:
dlc_folder = "/mnt/g/Shared drives/llorente-lab/Moseq/AnalyzedData/Revision-NORT/Revision-NORT-Viewable"
bbox_assignments_dir = "/mnt/g/Shared drives/llorente-lab/Moseq/AnalyzedData/Revision-NORT/bbox_assignments"

all_videos = [f for f in os.listdir(dlc_folder) if f.endswith('.mp4')]
novel_videos = [v for v in all_videos if 'NOVEL' in v.upper()]

# 4. Select NOVEL Videos

Filter the video directory to NOVEL-phase recordings only (filenames containing `'NOVEL'`). Only the NOVEL phase is scored — the FAMILIAR phase is used for habituation and is not relevant to the discrimination index.

In [9]:
def process_and_save_novel_videos(novel_videos, svm_model, scaler, dlc_folder, bbox_assignments_dir):
    """Process all NOVEL videos and save results"""
    
    results_dir = "/mnt/g/Shared drives/llorente-lab/Moseq/AnalyzedData/Revision-NORT/results_nort"
    os.makedirs(results_dir, exist_ok=True)
    
    for video_file in tqdm(novel_videos, desc="Processing NOVEL videos"):
        video_path = os.path.join(dlc_folder, video_file)
        video_stem = Path(video_file).stem
        
        # Create subfolder for this video
        video_results_dir = os.path.join(results_dir, video_stem)
        os.makedirs(video_results_dir, exist_ok=True)
        
        # Predict
        result = predict_with_object_assignment(video_path, svm_model, scaler, dlc_folder, bbox_assignments_dir)
        
        if result is None:
            print(f"Skipped {video_stem}")
            continue
        
        # Save JSON results
        json_output = os.path.join(video_results_dir, f"{video_stem}_predictions.json")
        with open(json_output, 'w') as f:
            json.dump(result, f, indent=2)
        
        # Create labeled video
        video_output = os.path.join(video_results_dir, f"{video_stem}_labeled.mp4")
        create_labeled_video(video_path, result['frame_results'], video_output)
        
        print(f"Completed {video_stem}")

print("Processing all NOVEL videos...")
process_and_save_novel_videos(novel_videos, svm_optimized, scaler, dlc_folder, bbox_assignments_dir)
print("\nAll NOVEL videos processed!")

Processing all NOVEL videos...


Processing NOVEL videos:   1%|▎                                      | 1/140 [00:22<52:59, 22.87s/it]

Completed VK_20251026_a_BSL_NOVEL_20251029133242_ir_viewable


Processing NOVEL videos:   1%|▌                                    | 2/140 [00:55<1:06:05, 28.74s/it]

Completed VK_20251026_e_BSL_NOVEL_20251029132956_ir_viewable


Processing NOVEL videos:   2%|▊                                    | 3/140 [01:24<1:05:45, 28.80s/it]

Completed VK_20251026_b_BSL_NOVEL_20251029133840_ir_viewable


Processing NOVEL videos:   3%|█                                    | 4/140 [01:55<1:07:13, 29.66s/it]

Completed VK_20251026_f_BSL_NOVEL_20251029133551_ir_viewable


Processing NOVEL videos:   4%|█▎                                   | 5/140 [02:26<1:07:27, 29.98s/it]

Completed VK_20251026_c_BSL_NOVEL_20251029134421_ir_viewable


Processing NOVEL videos:   4%|█▌                                   | 6/140 [02:52<1:04:03, 28.68s/it]

Completed VK_20251026_g_BSL_NOVEL_20251029134142_ir_viewable


Processing NOVEL videos:   5%|█▉                                     | 7/140 [03:15<59:58, 27.06s/it]

Completed VK_20251026_d_BSL_NOVEL_20251029135021_ir_viewable


Processing NOVEL videos:   6%|██▏                                    | 8/140 [03:42<58:54, 26.78s/it]

Completed VK_20251026_h_BSL_NOVEL_20251029134742_ir_viewable


Processing NOVEL videos:   6%|██▌                                    | 9/140 [04:05<56:25, 25.84s/it]

Completed VK_20251026_t_BSL_NOVEL_20251029154407_ir_viewable


Processing NOVEL videos:   7%|██▌                                 | 10/140 [04:39<1:01:14, 28.27s/it]

Completed VK_20251026_m_BSL_NOVEL_20251029142631_ir_viewable


Processing NOVEL videos:   8%|██▉                                   | 11/140 [05:04<58:13, 27.08s/it]

Completed VK_20251026_j_BSL_NOVEL_20251029143402_ir_viewable


Processing NOVEL videos:   9%|███▎                                  | 12/140 [05:29<56:34, 26.52s/it]

Completed VK_20251026_n_BSL_NOVEL_20251029143253_ir_viewable


Processing NOVEL videos:   9%|███▌                                  | 13/140 [05:52<54:17, 25.65s/it]

Completed VK_20251026_k_BSL_NOVEL_20251029144015_ir_viewable


Processing NOVEL videos:  10%|███▊                                  | 14/140 [06:18<53:43, 25.58s/it]

Completed VK_20251026_o_BSL_NOVEL_20251029143847_ir_viewable


Processing NOVEL videos:  11%|████                                  | 15/140 [06:52<58:38, 28.15s/it]

Completed VK_20251026_l_BSL_NOVEL_20251029144605_ir_viewable


Processing NOVEL videos:  11%|████▎                                 | 16/140 [07:18<56:38, 27.41s/it]

Completed VK_20251026_p_BSL_NOVEL_20251029144522_ir_viewable


Processing NOVEL videos:  12%|████▌                                 | 17/140 [07:41<53:56, 26.31s/it]

Completed VK_20251026_q_BSL_NOVEL_20251029152611_ir_viewable


Processing NOVEL videos:  13%|████▉                                 | 18/140 [08:07<52:48, 25.97s/it]

Completed VK_20251026_u_BSL_NOVEL_20251029152446_ir_viewable


Processing NOVEL videos:  14%|█████▏                                | 19/140 [08:38<55:44, 27.64s/it]

Completed VK_20251026_r_BSL_NOVEL_20251029153204_ir_viewable


Processing NOVEL videos:  14%|█████▍                                | 20/140 [09:04<54:00, 27.00s/it]

Completed VK_20251026_v_BSL_NOVEL_20251029153039_ir_viewable


Processing NOVEL videos:  15%|█████▋                                | 21/140 [09:26<50:59, 25.71s/it]

Completed VK_20251026_s_BSL_NOVEL_20251029153748_ir_viewable


Processing NOVEL videos:  16%|█████▉                                | 22/140 [09:52<50:29, 25.67s/it]

Completed VK_20251026_w_BSL_NOVEL_20251029153629_ir_viewable


Processing NOVEL videos:  16%|██████▏                               | 23/140 [10:18<50:04, 25.68s/it]

Completed VK_20251026_x_BSL_NOVEL_20251029154213_ir_viewable


Processing NOVEL videos:  17%|██████▌                               | 24/140 [10:50<53:46, 27.81s/it]

Completed VK_20251026_y_BSL_NOVEL_20251029162140_ir_viewable


Processing NOVEL videos:  18%|██████▊                               | 25/140 [11:16<51:56, 27.10s/it]

Completed VK_20251026_ac_BSL_NOVEL_20251029162028_ir_viewable


Processing NOVEL videos:  19%|███████                               | 26/140 [11:39<49:20, 25.97s/it]

Completed VK_20251026_z_BSL_NOVEL_20251029162742_ir_viewable


Processing NOVEL videos:  19%|███████▎                              | 27/140 [12:09<51:02, 27.11s/it]

Completed VK_20251026_ad_BSL_NOVEL_20251029162625_ir_viewable


Processing NOVEL videos:  20%|███████▌                              | 28/140 [12:38<51:34, 27.63s/it]

Completed VK_20251026_aa_BSL_NOVEL_20251029163328_ir_viewable


Processing NOVEL videos:  21%|███████▊                              | 29/140 [13:05<51:07, 27.63s/it]

Completed VK_20251026_ae_BSL_NOVEL_20251029163208_ir_viewable


Processing NOVEL videos:  21%|████████▏                             | 30/140 [13:29<48:38, 26.53s/it]

Completed VK_20251026_ab_BSL_NOVEL_20251029163926_ir_viewable


Processing NOVEL videos:  22%|████████▍                             | 31/140 [13:54<47:25, 26.10s/it]

Completed VK_20251026_af_BSL_NOVEL_20251029163756_ir_viewable


Processing NOVEL videos:  23%|████████▋                             | 32/140 [14:25<49:26, 27.47s/it]

Completed VK_20251026_ah_BSL_NOVEL_20251029165743_ir_viewable


Processing NOVEL videos:  24%|████████▉                             | 33/140 [14:49<46:58, 26.34s/it]

Completed VIK_20251026_ag_BSL_NOVEL_20251029171530_ir_viewable


Processing NOVEL videos:  24%|█████████▏                            | 34/140 [15:12<44:52, 25.40s/it]

Completed VK_20251026_ah_BSL_NOVEL_20251029172114_ir_viewable


Processing NOVEL videos:  25%|█████████▌                            | 35/140 [15:35<43:14, 24.71s/it]

Completed VK_20251026_ai_BSL_NOVEL_20251029172716_ir_viewable


Processing NOVEL videos:  26%|█████████▊                            | 36/140 [15:59<42:18, 24.41s/it]

Completed VK_20251026_aj_BSL_NOVEL_20251029173310_ir_viewable


Processing NOVEL videos:  26%|██████████                            | 37/140 [16:29<44:42, 26.04s/it]

Completed VK_20251026_z_30DPI_NOVEL_20251202161646_ir_viewable


Processing NOVEL videos:  27%|██████████▎                           | 38/140 [16:55<44:11, 25.99s/it]

Completed VK_20251026_a_30DPI_NOVEL_20251202132421_ir_viewable


Processing NOVEL videos:  28%|██████████▌                           | 39/140 [17:19<42:45, 25.40s/it]

Completed VK_20251026_aa_30DPI_NOVEL_20251202162319_ir_viewable


Processing NOVEL videos:  29%|██████████▊                           | 40/140 [17:43<41:35, 24.96s/it]

Completed VK_20251026_ab_30DPI_NOVEL_20251202162942_ir_viewable


Processing NOVEL videos:  29%|███████████▏                          | 41/140 [18:07<40:48, 24.73s/it]

Completed VK_20251026_ac_30DPI_NOVEL_20251202160734_ir_viewable


Processing NOVEL videos:  30%|███████████▍                          | 42/140 [18:36<42:38, 26.10s/it]

Completed VK_20251026_ad_30DPI_NOVEL_20251202161429_ir_viewable


Processing NOVEL videos:  31%|███████████▋                          | 43/140 [19:02<42:10, 26.09s/it]

Completed VK_20251026_ae_30DPI_NOVEL_20251202162114_ir_viewable


Processing NOVEL videos:  31%|███████████▉                          | 44/140 [19:26<40:53, 25.56s/it]

Completed VK_20251026_af_30DPI_NOVEL_20251202162801_ir_viewable


Processing NOVEL videos:  32%|████████████▏                         | 45/140 [19:50<39:25, 24.90s/it]

Completed VK_20251026_ag_30DPI_NOVEL_20251202170822_ir_viewable


Processing NOVEL videos:  33%|████████████▍                         | 46/140 [20:13<38:06, 24.33s/it]

Completed VK_20251026_ah_30DPI_NOVEL_20251202171443_ir_viewable


Processing NOVEL videos:  34%|████████████▊                         | 47/140 [20:44<41:07, 26.53s/it]

Completed VK_20251026_ai_30DPI_NOVEL_20251202172121_ir_viewable


Processing NOVEL videos:  34%|█████████████                         | 48/140 [21:09<39:55, 26.04s/it]

Completed VK_20251026_aj_30DPI_NOVEL_20251202172744_ir_viewable


Processing NOVEL videos:  35%|█████████████▎                        | 49/140 [21:33<38:11, 25.18s/it]

Completed VK_20251026_b_30DPI_NOVEL_20251202133113_ir_viewable


Processing NOVEL videos:  36%|█████████████▌                        | 50/140 [21:56<36:58, 24.65s/it]

Completed VK_20251026_c_30DPI_NOVEL_20251202133735_ir_viewable


Processing NOVEL videos:  36%|█████████████▊                        | 51/140 [22:20<36:25, 24.56s/it]

Completed VK_20251026_d_30DPI_NOVEL_20251202134411_ir_viewable


Processing NOVEL videos:  37%|██████████████                        | 52/140 [23:16<49:47, 33.95s/it]

Completed VK_20251026_e_30DPI_NOVEL_20251202132122_ir_viewable


Processing NOVEL videos:  38%|██████████████▍                       | 53/140 [23:41<45:24, 31.32s/it]

Completed VK_20251026_f_30DPI_NOVEL_20251202132809_ir_viewable


Processing NOVEL videos:  39%|██████████████▋                       | 54/140 [24:06<41:56, 29.27s/it]

Completed VK_20251026_g_30DPI_NOVEL_20251202133508_ir_viewable


Processing NOVEL videos:  39%|██████████████▉                       | 55/140 [24:33<40:34, 28.65s/it]

Completed VK_20251026_h_30DPI_NOVEL_20251202134139_ir_viewable


Processing NOVEL videos:  40%|███████████████▏                      | 56/140 [25:00<39:33, 28.25s/it]

Completed VK_20251026_i_30DPI_NOVEL_20251202142406_ir_viewable


Processing NOVEL videos:  41%|███████████████▍                      | 57/140 [25:24<37:20, 26.99s/it]

Completed VK_20251026_j_30DPI_NOVEL_20251202143029_ir_viewable


Processing NOVEL videos:  41%|███████████████▋                      | 58/140 [25:49<36:00, 26.35s/it]

Completed VK_20251026_k_30DPI_NOVEL_20251202143705_ir_viewable


Processing NOVEL videos:  42%|████████████████                      | 59/140 [26:16<35:42, 26.45s/it]

Completed VK_20251026_l_30DPI_NOVEL_20251202144332_ir_viewable


Processing NOVEL videos:  43%|████████████████▎                     | 60/140 [26:44<35:44, 26.80s/it]

Completed VK_20251026_m_30DPI_NOVEL_20251202142211_ir_viewable


Processing NOVEL videos:  44%|████████████████▌                     | 61/140 [27:16<37:36, 28.57s/it]

Completed VK_20251026_n_30DPI_NOVEL_20251202142840_ir_viewable


Processing NOVEL videos:  44%|████████████████▊                     | 62/140 [27:42<36:05, 27.76s/it]

Completed VK_20251026_o_30DPI_NOVEL_20251202143503_ir_viewable


Processing NOVEL videos:  45%|█████████████████                     | 63/140 [28:10<35:42, 27.83s/it]

Completed VK_20251026_p_30DPI_NOVEL_20251202144129_ir_viewable


Processing NOVEL videos:  46%|█████████████████▎                    | 64/140 [28:43<37:03, 29.26s/it]

Completed VK_20251026_r_30DPI_NOVEL_20251202151637_ir_viewable


Processing NOVEL videos:  46%|█████████████████▋                    | 65/140 [29:07<34:51, 27.89s/it]

Completed VK_20251026_s_30DPI_NOVEL_20251202152336_ir_viewable


Processing NOVEL videos:  47%|█████████████████▉                    | 66/140 [29:32<33:10, 26.90s/it]

Completed VK_20251026_t_30DPI_NOVEL_20251202153001_ir_viewable


Processing NOVEL videos:  48%|██████████████████▏                   | 67/140 [29:58<32:16, 26.52s/it]

Completed VK_20251026_u_30DPI_NOVEL_20251202151439_ir_viewable


Processing NOVEL videos:  49%|██████████████████▍                   | 68/140 [30:31<34:27, 28.71s/it]

Completed VK_20251026_v_30DPI_NOVEL_20251202152109_ir_viewable


Processing NOVEL videos:  49%|██████████████████▋                   | 69/140 [30:57<32:47, 27.71s/it]

Completed VK_20251026_w_30DPI_NOVEL_20251202152740_ir_viewable


Processing NOVEL videos:  51%|███████████████████▎                  | 71/140 [31:22<21:37, 18.81s/it]

Completed VK_20251026_y_30DPI_NOVEL_20251202161036_ir_viewable
Missing bbox for VK_20251026_v_30DPI_NOVEL_20251202152109_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_v_30DPI_NOVEL_20251202152109_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_t_BSL_NOVEL_20251029154407_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_t_BSL_NOVEL_20251029154407_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_k_30DPI_NOVEL_20251202143705_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_k_30DPI_NOVEL_20251202143705_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_u_30DPI_NOVEL_20251202151439_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_lab

Processing NOVEL videos:  66%|████████████████████████▉             | 92/140 [31:22<01:12,  1.50s/it]

Missing bbox for VK_20251026_c_BSL_NOVEL_20251029134421_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_c_BSL_NOVEL_20251029134421_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_d_BSL_NOVEL_20251029135021_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_d_BSL_NOVEL_20251029135021_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_q_BSL_NOVEL_20251029152611_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_q_BSL_NOVEL_20251029152611_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_ac_30DPI_NOVEL_20251202160734_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_ac_30DPI_NOVEL_20251202160734_ir_viewableDLC_

Processing NOVEL videos:  75%|███████████████████████████▊         | 105/140 [31:22<00:27,  1.26it/s]

Missing bbox for VK_20251026_j_30DPI_NOVEL_20251202143029_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_j_30DPI_NOVEL_20251202143029_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_f_BSL_NOVEL_20251029133551_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_f_BSL_NOVEL_20251029133551_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_aj_BSL_NOVEL_20251029173310_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_aj_BSL_NOVEL_20251029173310_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_i_30DPI_NOVEL_20251202142406_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_i_30DPI_NOVEL_20251202142406_ir_viewable

Processing NOVEL videos:  93%|██████████████████████████████████▎  | 130/140 [31:22<00:03,  3.23it/s]

Missing bbox for VK_20251026_n_30DPI_NOVEL_20251202142840_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_n_30DPI_NOVEL_20251202142840_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_g_30DPI_NOVEL_20251202133508_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_g_30DPI_NOVEL_20251202133508_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_w_30DPI_NOVEL_20251202152740_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_w_30DPI_NOVEL_20251202152740_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_ae_30DPI_NOVEL_20251202162114_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_ae_30DPI_NOVEL_20251202162114_ir_

Processing NOVEL videos: 100%|█████████████████████████████████████| 140/140 [31:22<00:00, 13.45s/it]

Missing bbox for VK_20251026_t_30DPI_NOVEL_20251202153001_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_t_30DPI_NOVEL_20251202153001_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Missing bbox for VK_20251026_ad_BSL_NOVEL_20251029162625_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled
Skipped VK_20251026_ad_BSL_NOVEL_20251029162625_ir_viewableDLC_Resnet101_NORTDec17shuffle2_snapshot_best-90_filtered_p60_labeled

All NOVEL videos processed!


In [10]:
# List all bbox files
bbox_files = sorted([f for f in os.listdir(bbox_assignments_dir) if f.endswith('.pkl')])

print(f"Total bbox files: {len(bbox_files)}")
print(f"\nFirst 10 bbox files:")
for i, f in enumerate(bbox_files[:10]):
    print(f"  {i+1}. {f}")

# Check NOVEL videos that are missing bboxes
novel_video_stems = [Path(v).stem for v in novel_videos]

missing_bbox = []
has_bbox = []

for video_stem in novel_video_stems:
    expected_bbox = f"{video_stem}_bbox_assignments.pkl"
    if expected_bbox in bbox_files:
        has_bbox.append(video_stem)
    else:
        missing_bbox.append(video_stem)

print(f"\n{'='*80}")
print(f"NOVEL videos WITH bbox: {len(has_bbox)}")
print(f"NOVEL videos MISSING bbox: {len(missing_bbox)}")

if len(missing_bbox) > 0:
    print(f"\nFirst 5 missing:")
    for i, v in enumerate(missing_bbox[:5]):
        print(f"  {i+1}. {v}")

Total bbox files: 137

First 10 bbox files:
  1. VIK_20251026_ag_BSL_NOVEL_20251029171530_ir_viewable_bbox_assignments.pkl
  2. VK_20251025_ag_30DPI_FAMILIAR_20251202164224_ir_viewable_bbox_assignments.pkl
  3. VK_20251025_ai_30DPI_FAMILIAR_20251202165520_ir_viewable_bbox_assignments.pkl
  4. VK_20251026_a_30DPI_FAMILIAR_20251202125804_ir_viewable_bbox_assignments.pkl
  5. VK_20251026_a_30DPI_NOVEL_20251202132421_ir_viewable_bbox_assignments.pkl
  6. VK_20251026_a_BSL_FAMILIAR_20251029130558_ir_viewable_bbox_assignments.pkl
  7. VK_20251026_a_BSL_NOVEL_20251029133242_ir_viewable_bbox_assignments.pkl
  8. VK_20251026_aa_30DPI_NOVEL_20251202162319_ir_viewable_bbox_assignments.pkl
  9. VK_20251026_aa_BSL_FAMILIAR_20251029160855_ir_viewable_bbox_assignments.pkl
  10. VK_20251026_aa_BSL_NOVEL_20251029163328_ir_viewable_bbox_assignments.pkl

NOVEL videos WITH bbox: 70
NOVEL videos MISSING bbox: 70

First 5 missing:
  1. VK_20251026_v_30DPI_NOVEL_20251202152109_ir_viewableDLC_Resnet101_NORTDe

# 5. Debug: Bbox File Matching

Half the NOVEL videos appear to be missing bbox files when matched by exact filename stem. The issue is that some video filenames include a DLC suffix (e.g., `...DLC_Resnet101_NORTDec17shuffle2_..._labeled`) appended to the base video name. The bbox files were saved using only the base stem before the `DLC` suffix. The next cell fixes this by stripping the suffix before lookup.

In [11]:
# Get NOVEL videos that actually have bbox files
novel_videos_with_bbox = []

for video_file in novel_videos:
    video_stem = Path(video_file).stem
    
    # Try exact match first
    expected_bbox = f"{video_stem}_bbox_assignments.pkl"
    if expected_bbox in bbox_files:
        novel_videos_with_bbox.append(video_file)
    else:
        # Try stripping DLC suffix
        if 'DLC' in video_stem:
            base_stem = video_stem.split('DLC')[0]
            expected_bbox = f"{base_stem}_bbox_assignments.pkl"
            if expected_bbox in bbox_files:
                novel_videos_with_bbox.append(video_file)

print(f"NOVEL videos with bbox files: {len(novel_videos_with_bbox)} / {len(novel_videos)}")
print(f"\nProcessing only these {len(novel_videos_with_bbox)} videos...")

NOVEL videos with bbox files: 140 / 140

Processing only these 140 videos...


In [12]:
def predict_with_object_assignment_fixed(video_path, svm_model, scaler, dlc_folder, bbox_assignments_dir):
    """Predict exploratory behavior per frame, handling DLC suffix in video names"""
    
    video_stem = Path(video_path).stem
    
    # Handle DLC suffix in video name
    bbox_stem = video_stem.split('DLC')[0] if 'DLC' in video_stem else video_stem
    bbox_file = os.path.join(bbox_assignments_dir, f"{bbox_stem}_bbox_assignments.pkl")
    
    if not os.path.exists(bbox_file):
        print(f"Missing bbox for {video_stem}")
        return None
    
    with open(bbox_file, 'rb') as f:
        bbox_data = pickle.load(f)
    
    # Load DLC
    try:
        dlc_csv = find_dlc_csv(video_path, dlc_folder)
        dlc_data = load_dlc_csv(dlc_csv)
    except Exception as e:
        print(f"Error loading DLC for {video_stem}: {e}")
        return None
    
    n_frames = len(dlc_data['nose_x'])
    
    # Get object names
    first_frame_bboxes = bbox_data[0]
    object_names = sorted(first_frame_bboxes.keys())
    
    # Results
    frame_results = []
    
    for frame_idx in range(n_frames):
        if frame_idx not in bbox_data:
            frame_results.append({
                'frame_idx': frame_idx,
                'exploratory': False,
                'probability': 0.0,
                'exploring_object': None,
                'distance_to_objects': {}
            })
            continue
        
        bboxes = bbox_data[frame_idx]
        
        nose_x = dlc_data['nose_x'][frame_idx]
        nose_y = dlc_data['nose_y'][frame_idx]
        tail_x = dlc_data['tail_x'][frame_idx]
        tail_y = dlc_data['tail_y'][frame_idx]
        
        # Calculate features for ALL objects (16 features)
        all_features = []
        for obj_name in sorted(bboxes.keys()):
            vertices = bbox_to_vertices(bboxes[obj_name])
            features = calculate_features_for_frame(nose_x, nose_y, tail_x, tail_y, vertices)
            all_features.extend(features)
        
        all_features = np.array(all_features).reshape(1, -1)
        
        if np.any(np.isnan(all_features)):
            frame_results.append({
                'frame_idx': frame_idx,
                'exploratory': False,
                'probability': 0.0,
                'exploring_object': None,
                'distance_to_objects': {}
            })
            continue
        
        # Predict exploratory behavior for this frame
        features_scaled = scaler.transform(all_features)
        pred = svm_model.predict(features_scaled)[0]
        prob = svm_model.predict_proba(features_scaled)[0, 1]
        
        # Calculate distance to each object
        distances = {}
        for obj_name, bbox in bboxes.items():
            obj_center_x = (bbox['xmin'] + bbox['xmax']) / 2
            obj_center_y = (bbox['ymin'] + bbox['ymax']) / 2
            dist = np.sqrt((nose_x - obj_center_x)**2 + (nose_y - obj_center_y)**2)
            distances[obj_name] = dist
        
        # If exploratory, assign to closest object
        exploring_object = None
        if pred == 1:
            exploring_object = min(distances, key=distances.get)
        
        frame_results.append({
            'frame_idx': frame_idx,
            'exploratory': bool(pred),
            'probability': float(prob),
            'exploring_object': exploring_object,
            'distance_to_objects': distances
        })
    
    return {
        'video_stem': video_stem,
        'n_frames': n_frames,
        'object_names': object_names,
        'frame_results': frame_results
    }

print("Updated prediction function loaded!")

Updated prediction function loaded!


# 6. Fixed Inference Function and Final Run

`predict_with_object_assignment_fixed` is identical to the original function but strips the DLC suffix from the video stem before looking up the bbox file. This resolves the missing-bbox issue and allows all 140 NOVEL videos to be processed.

The final `process_and_save_novel_videos` call runs inference on all videos and writes:
- `<video_stem>_predictions.json` — frame-level predictions, probabilities, assigned object, distances
- `<video_stem>_labeled.mp4` — annotated video with overlaid exploratory/non-exploratory labels

In [ ]:
def process_and_save_novel_videos(novel_videos, svm_model, scaler, dlc_folder, bbox_assignments_dir):
    """Process all NOVEL videos and save results"""
    
    results_dir = "/mnt/g/Shared drives/llorente-lab/Moseq/AnalyzedData/Revision-NORT/results_nort"
    os.makedirs(results_dir, exist_ok=True)
    
    for video_file in tqdm(novel_videos, desc="Processing NOVEL videos"):
        video_path = os.path.join(dlc_folder, video_file)
        video_stem = Path(video_file).stem
        
        # Create subfolder for this video
        video_results_dir = os.path.join(results_dir, video_stem)
        os.makedirs(video_results_dir, exist_ok=True)
        
        # Predict
        result = predict_with_object_assignment_fixed(video_path, svm_model, scaler, dlc_folder, bbox_assignments_dir)
        
        if result is None:
            print(f"Skipped {video_stem}")
            continue
        
        # Save JSON results
        json_output = os.path.join(video_results_dir, f"{video_stem}_predictions.json")
        with open(json_output, 'w') as f:
            json.dump(result, f, indent=2)
        
        # Create labeled video
        video_output = os.path.join(video_results_dir, f"{video_stem}_labeled.mp4")
        create_labeled_video(video_path, result['frame_results'], video_output)
        
        print(f"Completed {video_stem}")

print("Processing all NOVEL videos...")
process_and_save_novel_videos(novel_videos, svm_optimized, scaler, dlc_folder, bbox_assignments_dir)
print("\nAll NOVEL videos processed!")